lib/irb.rb

```diff
@@ -562,7 +562,8 @@ def eval_input
             is_assignment = assignment_expression?(line)
             evaluate_line(line, line_no)
 
-            if @context.echo?
+            # Don't echo if the line ends with a semicolon
+            if @context.echo? && !line.match?(/;\s*\z/)
```

Comment:
I think this simple implementation is good enough.

However, I found a false positive. Using`tokens.last` will be more strict. What do you think?
```ruby
puts %;string;
```
```ruby
puts +% ; 
```
```ruby
puts <<';'
;
```
```ruby
puts <<' '
heredoc;
 
```

And maybe a false negative, maybe not. (Checking this will be very hard)
```ruby
puts(<<A);
A
```

Reply:
Great catch! And yeah I agree it's better to make this check with `Ripper`. But I'd like to avoid invoking it outside of `RubyLex` and want to make this property an yield argument of `each_top_level_statement` instead (like #670). WDYT?